### ISSUE TITLE
removing discard from the fragment shader significantly improves performance on iOS

### ISSUE THREAD

**Issue by wtholliday** [memononen/nanovg#214]
Before:

![before](https://cloud.githubusercontent.com/assets/841144/6924316/9a90b34a-d78c-11e4-8d4c-42c56cf37d87.png)

After:

![after](https://cloud.githubusercontent.com/assets/841144/6924323/a1eb90c4-d78c-11e4-8a6a-4afa667b9938.png)

I did notice some rendering artifacts for self-intersecting paths:

![img_0052](https://cloud.githubusercontent.com/assets/841144/6924440/64414fec-d78d-11e4-88a9-21786d70744b.PNG)

vs.

![img_0053](https://cloud.githubusercontent.com/assets/841144/6924446/6bf1c3a2-d78d-11e4-9baa-a89befa7e96e.PNG)

How should I fix the artifacts? Should there be an option exposed to turn off discard (it's a no-no on iOS)?

**Comment by dougbinks** [comment]
It should be possible to set the alpha value to 0.0 instead of using discard to make the pixel completely transparent. Discard is indeed an issue on many mobile configurations, so this should likely be done for all OpenGL ES configurations.

**Comment by wtholliday** [comment]
@dougbinks, surprisingly, setting the alpha to 0.0 doesn't fix the artifacts. In fact it creates a new problem with strokes:

![img_0054](https://cloud.githubusercontent.com/assets/841144/6925681/e66f80da-d795-11e4-84b0-d9d34a8747b8.PNG)

I did the following:

```
    "#ifdef EDGE_AA\n"
    "	if (strokeAlpha < strokeThr) { result.a = 0.0; };\n"
    "#endif\n"
```

**Comment by dougbinks** [comment]
The blend function is glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA); so this is likely due to depth or stencil testing issues and so not easily solvable though Mikko might be able to come up with a solution.

Since the discard is only on the EDGE_AA path, I would not use the NVG_ANTIALIAS if you want performance.

**Comment by wtholliday** [comment]
Good thought about the stenciling. Turning off NVG_STENCIL_STROKES gets rid of the artifacts, but of course then self-intersecting paths are drawn differently (which I can live with).

**Comment by memononen** [comment]
Good catch! I wonder if splitting the shaders would make it faster on mobile too? So that if one shader uses discard, all shaders don't pay the penalty?

The idea of the discard is that we first draw the "fill" of the stroke, but only once, and then on second pass we draw the AA bit of the stroke. It either needs two sets of geometry, or discard/alphatest. I chose to use discard.

**Comment by dougbinks** [comment]
Splitting the shaders may help, since the discard is driven by data it can't be optimized out and tile based GPUs then pay a big penalty (there's a cost too on other GPUs, but it's far less).

**Comment by wtholliday** [comment]
If two AA parts of a path each mostly cover a pixel, would the color be computed as if coverage is greater than the area of the pixel? I'm trying to understand the existing implementation.

**Comment by memononen** [comment]
The AA in nanovg works so that first the solid (non-AA) part of the shape is drawn, creating a stencil mask of the fill of the shape. Then the solid shape is filled using a quad. Finally, the stencil test is inverted, and an antialiased line is drawn around the perimeter (since AA-line drawing is iffy, this line is created as a triangle strip). The stencil mask is also used to handle the fill rule of the shape so that you can have holes.

The `NVG_STENCIL_STROKES` works similarly, but avoids the fill pass by drawing the solid shape and filling in the same pass. Essentially it uses the stencil buffer to only draw once at a given location. On the perimeter pass the stencil test is inverted and the AA edge of the stroke is drawn. The stroke geometry is build so that one segment of the path contains two triangles. The `strokeThr` is calculated so that it culls out the AA portion of the stroke during the solid pass.

That means that with both methods, the AA edge (I call it AA-fringe) can be slightly wrong, as it will be overdrawn in case the shape intersects or overaps. You can see this sometimes especially with light colors, but generally it works well. You can also configure NanoVG to not generate any of the AA fringes and set your frame buffer to use multisampling AA (i.e. MSAA) if correct coverage is required.

**Comment by wtholliday** [comment]
I think you could draw the entire stroke in a single pass (with stenciling). Instead of overdraw of self-intersecting AA edges, you'd have underdraw. Would that work?

**Comment by memononen** [comment]
If you draw the entire stroke with stencil in single pass you'll get the artifacts as in your first screenshots. That light line is the anti-aliased bit of the vertical segment, which happens to be drawn first.

**Comment by wtholliday** [comment]
Ah, of course!

Well, at least it seems like the discard isn't needed when `NVG_STENCIL_STROKES` is off, right?

**Comment by memononen** [comment]
That's correct.
